### PR TITLE
[api] Move `Names` out of API.

### PR DESCRIPTION
**Pull request by ejgallego**
This is a proof of concept in order to do the third iteration of the
"API" design.

The idea is provide in each module needing it a `Public` and
`Internal` system. Then, we make the public part available to
consumers.

This provides significant advantages wrt previous approach, at a
non-so-modest cost of having to update the base implementation
itself. However, the transition could be implemented gradually.

c.f. issue #6008

### CODE REVIEW

File: kernel/names.mli [hunk 1]
```diff
@@ -737,100 +741,108 @@ end
...
+[@@ocaml.deprecated "Same as [MutInd.debug_print]"]
+
+val debug_string_of_mind : MutInd.t -> string
+[@@ocaml.deprecated "Same as [MutInd.debug_to_string]"]
+
+end
+
+(* Nothing is internal for now *)
+module Internal : sig
+end
```
Comment (SkySkimmer) [inline kernel/names.mli]: This makes it impossible to judge if the idea of having an Internal module will work.

Comment (ejgallego) [reply kernel/names.mli]: Well, what should we put there? AFAICS the whole of Names should be exported. We could do large refactoring in `Names` using `module type`s and indeed export a richer system in `Internal` with things such as `hash`, but yeah, for now we must move on.

**Comment by ejgallego** [comment]
Ok so I am gonna stop here until we get some more feedback and discuss the options available. In particular, it looks like given a module `A` such that:

```ocaml
(* A.mli *)
module Public : sig
   type t = ...
end
include module type of struct include Public end
module Internal : sig
end
```
is the most sensible option, as this will make `A.t = A.Public.t`, right? However, a different issue is what do we do then with API.mli, what should we expose there in the "transition phase"?

**Comment by ejgallego** [comment]
A second question is the idiom used in `A.ml`

```ocaml
(* a.ml *)
module Internal = struct
....
end
module Public = Internal
include Internal
```
I am far from an OCaml expert so I wonder if this is going to cause some issues. Maybe @gasche has some advice here?

**Review by herbelin — COMMENTED** [review]
This look good to me.

About the `Constant.t`, I already expressed a view and I think that there are people more opiniated than me on this subject, so I have nothing special more to say.

Otherwise, an idea, but maybe it was already proposed, I don't remember.

Would it make sense that, instead of duplicating code (as in the current `API`), or divising code into a public and private part (as here), that each `val` of each `mli` comes with a mini comment. I don't know if user-level `@` tags exists, but we could simply have a light comment, say `(*E*)` in front or after each `val`, so that one (or more API) files could be automatically processed from the existing `mli` files.

One could have several flags, like `(*I*)` for a function to be used in the non-plugin part, `(*S*)` for a function to be used only in the current subsystem, etc.

One could imagine other flags like `(*DE*)` for exported for compatibility but discouraged, and so on. The main idea being that the control is done at the level of the `mli` file where the function is defined.

**Comment by SkySkimmer** [comment]
@ejgallego I think you should merge the other PRs into this one, I don't think there's much point in separating discussion about Names/UGraph/etc.

**Comment by SkySkimmer** [comment]
@herbelin that sounds cool but I don't know how to implement it.

**Comment by ppedrot** [comment]
@herbelin This was more or less the approach I was advocating for. I think it's possible to do that with a PPX, but it seems it is not yet a safe bet due to portability across version...

**Comment by ejgallego** [comment]
> Would it make sense that, instead of duplicating code (as in the current API), or divising code into a public and private part (as here), that each val of each mli comes with a mini comment. I don't know if user-level @ tags exists, but we could simply have a light comment, say (*E*) in front or after each val, so that one (or more API) files could be automatically processed from the existing mli files.
> 
> One could have several flags, like (*I*) for a function to be used in the non-plugin part, (*S*) for a function to be used only in the current subsystem, etc.
> 
> One could imagine other flags like (*DE*) for exported for compatibility but discouraged, and so on. The main idea being that the control is done at the level of the mli file where the function is defined.

Well, the control with this approach is done at the level of the `.mli` file, but only for visibility.

About the system you propose, I dunno, on paper it is a good idea but how are you going to enforce the properties?

Indeed you can add coq-specific attributes and use a ppx to do something. So in principle it could happen, using `[@@coq.public ]`, etc... Would such alternative exists, I would be happy to support it, but it looks like a lot of work both in design and implementation.

**Comment by ejgallego** [comment]
I think you folks are underestimating the cost of implementing such a system in a reasonable way, but I may be wrong of course.

**Comment by ejgallego** [comment]
> @ejgallego I think you should merge the other PRs into this one, I don't think there's much point in separating discussion about Names/UGraph/etc.

I don't think so, I'd like each module to be reviewed separately.

Moreover, there are many things to discuss, this approach has some problems. I don't understand what is going with the module system, see for instance https://discuss.ocaml.org/t/a-question-on-type-module-equality/1071

So, if I use `module Names  : module type of struct include Names.Public end` in API to avoid declaring explicit aliases in API, I start to get:
```ocaml
File "plugins/syntax/z_syntax.ml", line 30, characters 30-59:
Error: This expression has type API.Names.Id.t list
       but an expression was expected of type Names.Public.module_ident list
       Type API.Names.Id.t = Names.Public.Id.t is not compatible with type
         Names.Public.module_ident 
```

**Review by ppedrot — APPROVED** [review]
LGTM.

**Comment by ejgallego** [comment]
I'm closing this a for now we will only add `Internal` to certain modules and it turns out that `Names` has no private components.

**Pull request closed without merging**